src/eg/src/guardian_proof.rs

```diff
@@ -0,0 +1,276 @@
+use std::borrow::Borrow;
+
+use crate::{
+    election_parameters::ElectionParameters,
+    guardian::GuardianIndex,
+    guardian_public_key_info::GuardianPublicKeyInfo,
+    guardian_secret_key::GuardianSecretKey,
+    hash::{eg_h, HValue},
+};
+use anyhow::{ensure, Context, Ok, Result};
+use num_bigint::BigUint;
+use serde::{Deserialize, Serialize};
+use util::{csprng::Csprng, integer_util::to_be_bytes_left_pad};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PossessionProofChallenge(
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub BigUint,
+);
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PossessionProofResponse(
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub BigUint,
+);
+
+/// Proof of Possession for the coefficient in a guardian's secret key.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GuardianProof {
+    /// Guardian number, 1 <= i <= [`crate::varying_parameters::VaryingParameters::n`].
+    pub i: GuardianIndex,
+    /// Challenge vector
+    pub c: Vec<PossessionProofChallenge>,
+    /// Response vector
+    pub v: Vec<PossessionProofResponse>,
+}
+
+impl GuardianProof {
+    /// This function computes the challenge for the coefficient NIZK as specified in Equation (12)
+    /// The arguments are
+    /// - h_p - the parameter base hash
+    /// - i - the guardian index
+    /// - j - the coefficient index
+    /// - capital_k_i_j - the coefficient commitment
+    /// - h_i_j - the commit message
+    fn challenge(
+        h_p: HValue,
+        i: u32,
+        j: u32,
+        capital_k_i_j: &BigUint,
+        h_i_j: &BigUint,
+    ) -> PossessionProofChallenge {
+        // v = 0x10 | b(i,4) | b(j,4) | b(capital_k_i_j,512) | b(h_i_j,j,512)
+        let mut v = vec![0x10];
+        v.extend_from_slice(i.to_be_bytes().as_slice());
+        v.extend_from_slice(j.to_be_bytes().as_slice());
```

Comment:
I guess the big-endian serialization is specified in the standard?

Reply:
Yes it is (cf. Sections `5.1`). Should we make this more explicit in the code?